#' Structural variant evidence files
#'
#' Classes to handle structural variant evidence files used in the GATK-SV
#' pipeline.
#'
#' @details
#' The supported files are almost identical to the ones generated by the GATK
#' tool
#' [CollectSVEvidence](https://gatk.broadinstitute.org/hc/en-us/articles/35967530268443-CollectSVEvidence-BETA)
#' with the main difference being that the files referenced here combine
#' evidence from multiple samples into a single file. All files must be
#' compressed with bgzip and indexed with tabix. Each file/evidence type has a
#' wrapper class that is used to open the file and extract the evidence for a
#' genomic region. Each construction function returns an object of a class with
#' the same name as the constructor.
#'
#' # Paired-end Reads
#'
#' This file contains discordant read pairs and the format is the same as the
#' single sample version, though the "sample name" field may contain multiple
#' samples. An object for this file type can be constructed with the
#' `pe_file()` function.
#'
#' # Split Reads
#'
#' This file contains split read clippings and the format is the same as the
#' single sample version, though the "sample name" field may contain multiple
#' samples. An object for this file type can be constructed with the
#' `sr_file()` function.
#'
#' # Read Depth
#'
#' This file contains binned read depth. The file contains three columns
#' indentifying the interval and then columns, one per sample, of the read
#' depth of reads that start in each interval. There is also a header line that
#' starts with "#Chr", "Start", and "End" in the first three columns and then
#' the sample name of each sample in the file. An object for this file type can
#' be constructed with the `rd_file()`. The read depth file class additionally
#' requires a information on the median read depth of each sample. See
#' [read_median_coverages()].
#'
#' @param path Path to the evidence file.
#' @param medians Named numeric vector of median coverage values of samples.
#' @returns An object for the evidence type.
#'
#' @examples
#' # PE evidence
#'
#' tf_pe_src <- tempfile()
#' tf_pe_dest <- tempfile(fileext = ".gz")
#' src_conn <- file(tf_pe_src, open = "wt")
#' cat("chr14\t10000\t+\tchr14\t20000\t+\tabc0000\n", file = tf_pe_src)
#' cat("chr14\t11000\t+\tchr14\t15000\t+\tabc0000\n", file = tf_pe_src)
#' cat("chr14\t11000\t+\tchr14\t16000\t+\tabc0001\n", file = tf_pe_src)
#' close(src_conn)
#' Rsamtools::bgzip(tf_pe_src, tf_pe_dest)
#' Rsamtools::indexTabix(tf_pe_dest, seq = 1, start = 2, end = 2)
#' pe <- pe_file(tf_pe_dest)
#'
#' # RD evidence
#'
#' tf_medians <- tempfile()
#' writeLines(c("abc0000\tabc0001", "30\t30"), tf_medians)
#' tf_rd_src <- tempfile()
#' tf_rd_dest <- tempfile(fileext = ".gz")
#' src_conn <- file(tf_rd_src, open = "wt")
#' cat("#Chr\tStart\tEnd\tabc0000\tabc0001\n", file = src_conn)
#' cat("chr5\t1000000\t1000100\t40\t41\n", file = src_conn)
#' cat("chr5\t1000100\t1000200\t39\t41\n", file = src_conn)
#' cat("chr5\t1000200\t1000300\t30\t35\n", file = src_conn)
#' close(src_conn)
#'
#' Rsamtools::bgzip(tf_rd_src, tf_rd_dest)
#' Rsamtools::indexTabix(tf_rd_dest, seq = 1, start = 2, end = 3, comment = "#", zeroBased = TRUE)
#' rd_medians <- read_median_coverages(tf_medians)
#' rd <- rd_file(tf_rd_dest, rd_medians)
#'
#' @name svevidencefiles
NULL

#' Create a new `svevidence` object
#'
#' This class represents SV evidence over a genomic region. It is assumed that
#' the evidence sources all have the same set of samples.
#'
#' @param contig A string. The contig containing the region.
#' @param start The start of the region (1-start).
#' @param end The end of the region (inclusive).
#' @param pe A [`pe_file`][svevidencefiles] object for the discordant read
#'   pairs evidence.
#' @param sr A [`sr_file`][svevidencefiles] object for the split read evidence.
#' @param rd A [`rd_file`][svevidencefiles] object for the read depth evidence.
#' @param svtype SV type of the variant.
#' @param pad Fraction of region to add as padding when extracting the evidence
#'   in the region. The minimum padded start is 1 and the maximum padded end
#'   is 536870912.
#' @returns A `svevidence` object.
#' @export
#'
#' @examples
#' pe_path <- system.file(
#'     "extdata",
#'     "example.PE.txt.gz",
#'     package = "gorilla",
#'     mustWork = TRUE
#' )
#' sr_path <- system.file(
#'     "extdata",
#'     "example.SR.txt.gz",
#'     package = "gorilla",
#'     mustWork = TRUE
#' )
#' rd_path <- system.file(
#'     "extdata",
#'     "example.RD.txt.gz",
#'     package = "gorilla",
#'     mustWork = TRUE
#' )
#' medians_path <- system.file(
#'     "extdata",
#'     "example_medianCov.txt",
#'     package = "gorilla",
#'     mustWork = TRUE
#' )
#' rd_medians <- read_median_coverages(medians_path)
#'
#' pe <- pe_file(pe_path)
#' sr <- sr_file(sr_path)
#' rd <- rd_file(rd_path, rd_medians)
#'
#' sv <- svevidence("chr16", 28743149, 28745149, pe, sr, rd, "DUP")
svevidence <- function(contig, start, end, pe, sr, rd, svtype, pad = 0) {
    new_svevidence(contig, start, end, pe, sr, rd, svtype, pad)
}

#' @rdname subset_samples
#' @export
subset_samples.svevidence <- function(x, samples) {
    sample_id <- NULL
    rstart <- NULL
    pos <- NULL

    pe <- subset_samples(x$pe, samples)
    sr <- subset_samples(x$sr, samples)
    rd <- subset_samples(x$rd, samples)

    structure(
        list(
            pe = pe,
            sr = sr,
            rd = rd,
            svtype = x$svtype,
            region = x$region
        ),
        class = "svevidence"
    )
}

new_svevidence <- function(contig, start, end, pe, sr, rd, svtype, pad) {
    stopifnot(is_string(contig))
    stopifnot(is_number(start))
    stopifnot(is_number(end))
    stopifnot(inherits(pe, "pe_file"))
    stopifnot(inherits(sr, "sr_file"))
    stopifnot(inherits(rd, "rd_file"))
    stopifnot(is_string(svtype))
    stopifnot(is_number(pad) && pad >= 0)

    pad_size <- ceiling((end - start + 1) * pad)
    qstart <- max(1, start - pad_size)
    qend <- min(TABIX_MAX_SEQLEN, end + pad_size)
    pe_mat <- query(pe, contig, qstart, qend)
    sr_mat <- query(sr, contig, qstart, qend)
    rd_mat <- query(rd, contig, qstart, qend)
    region <- list(
        contig = contig,
        start = start,
        end = end,
        qstart = qstart,
        qend = qend
    )

    structure(
        list(
            pe = pe_mat,
            sr = sr_mat,
            rd = rd_mat,
            svtype = svtype,
            region = region
        ),
        class = "svevidence"
    )
}
